submdspan: static slices

mhoemmen opened the pull request:
1. When `submdspan` is given a `tuple<integral_constant<T0, Value0>, integral_constant<T1, Value1>>` as a slice specifier, preserve the compile-time values in the returned `mdspan`'s `extents_type`.
2. When `submdspan` is given an `integral_constant<T, Val>` as a slice specifier, preserve the compile-time values in the returned `mdspan`'s `extents_type`.
 
Change (1) can turn a dynamic extent into a static extent.  For example, for a rank-3 `dextents<int, 3>` mdspan `m`, `submdspan(m, full_extent, full_extent, tuple{integral_constant<int, 1>{}, integral_constant<int, 3>{}})` has `extents<int, dynamic_extent, dynamic_extent, 2>`.  Since the mapping is templated on the `extents` type, a custom layout could bake all the extents information into the type as non-type template parameters, rather than relying on `constexpr` for compile-time evaluation.  This could be helpful when crossing a compilation unit boundary without link-time optimization, for example.

Change (2) only affects the offset computation.  It might still be helpful for specific custom layouts, though.  It's also an easy change that just imitates (1).

mhoemmen committed:
Add submdspan test with integral_constant slice

It passes, because integral_constant is convertible to size_t.
We verify this in the test.

mhoemmen committed:
__wrap_slice tuple of integral_constant

__wrap_slice can now take a tuple of two integral_constant.
It converts the tuple to a tuple of two size_t for now,
as the machinery for dealing with a __slice_wrap of tuple of two
integral_constant isn't in place yet.

mhoemmen committed:
submdspan of tuple of two integral_constant

... now compiles.  It doesn't turn the extents from compile-time
to run-time yet, though.

mhoemmen committed:
submdspan tuple integral_constant -> extents

If you give submdspan a slice that's a tuple of two
integral_constant<size_t, ...>, it now knows how to pass the two
compile-time values into the result's extents_type.

mhoemmen committed:
submdspan: tuple of mixed integral_constant

Generalize the previous commits to work with tuple of two
integral_constant with any integer types (including mixed).

mhoemmen committed:
Make a test 3-D just to exercise more cases

mhoemmen committed:
submdspan: slice integral_constant

If submdspan is given an integral_constant as a slice specifier,
pass it along into the offset computation as a compile-time value.

This is perhaps less useful than the tuple of two integral_constant
case, which can turn a dynamic extent into a static extent.  However,
it's not hard to do.

mhoemmen committed:
submdspan: Add more compile-time slice tests

In particular, make sure that we have rank-3 tests
that exercise the (full, tuple, full) and
(tuple, full, tuple) use cases.

Reduce code duplication and remove outdated comments.

crtrott reviewed (changes requested):
Can you add a test for 3D with the static slice happening in the middle, feels like first and last might be special, otherwise looks good.

mhoemmen commented:
@crtrott wrote:
> Can you add a test for 3D with the static slice happening in the middle, feels like first and last might be special, otherwise looks good.

Thanks Christian! : - )  I added more 3-D tests, including

* (tuple, full, tuple),
* (full, tuple, full),
* (full, full, index), and
* (full, index, full).

I also increased the variety of layouts exercised, and added 1-D tests.
